### PR TITLE
Workspaces

### PR DESCRIPTION
**Pull request by ahmedosman**
Hi Luke, We rebased with the current master branch and applied our feature changes. 

This pull request is for the Workspace feature only. We'll be submitting a pull request per feature. 

The workspace feature in summary: divide a single instance of the digits server into multiple workspaces (where data, models could be trained within the workspaces).  The only major change here is we've changed the hierarchy of the jobs directory to be jobs/$workspace_name/$job_name

Please let me know if you have any questions

Ahmed

### CODE REVIEW

**Comment by lukeyeager** [comment]
Thanks @ahmedosman (and @DESHRAJ)! This is much more review-able :smiley:

> The only major change here is we've changed the hierarchy of the jobs directory to be jobs/$workspace_name/$job_name

That's fine as long as we can provide some sort of upgrade path for converting old jobs to the new directory structure. Currently, I end up with hundreds of empty workspaces named for job_ids:

invalid-workspaces

And apparently DIGITS is trying to load all sub-folders as jobs:
```
[Errno 2] No such file or directory: '/raid/jobs/dev/20150415-180806-1984/train_db/status.pickle'
[Errno 2] No such file or directory: '/raid/jobs/dev/20150415-180806-1984/val_db/status.pickle'
[Errno 2] No such file or directory: '/raid/jobs/dev/20150415-180813-8ef7/train_db/status.pickle'
[Errno 2] No such file or directory: '/raid/jobs/dev/20150415-180813-8ef7/val_db/status.pickle'
[Errno 2] No such file or directory: '/raid/jobs/dev/20150416-103304-7604/train_db/status.pickle'
```
We'll need to find some way to address that. Maybe just checking in `scheduler.load_past_jobs()` to see if a `status.pickle` file exists in `$jobs_dir/$dirname` and moving the directory to `$jobs_dir/Default/$dirname` if so?

File: digits/views.py [hunk 1]
```diff
@@ -319,6 +336,62 @@ def path_autocomplete():
...
     return json.dumps(result)
 
+@app.route('/workspaces', methods=['GET'])
+def show_workspaces():
+    """
+    Lists the workspaces in DIGITS
+    """
+    jobs_dir = config_value('jobs_dir')
+    workspaces = next(os.walk(jobs_dir))[1]
+    print workspaces
```
Comment (lukeyeager) [inline digits/views.py]: Debug print statement

File: digits/model/images/classification/views.py [hunk 2]
```diff
@@ -473,6 +484,18 @@ def get_datasets():
         cmp=lambda x,y: cmp(y.id(), x.id())
         )
         ]
+    
+def get_datasets_workspace(workspace):
+        list_jobs =   scheduler.return_workspace_jobs(workspace)
+        for j in list_jobs:
+            print j.id()  , j.name()
```
Comment (lukeyeager) [inline digits/model/images/classification/views.py]: Debug print statement

**Comment by lukeyeager** [comment]
@jmancewicz will you take a look at this? Does this line up with how you were envisioning user spaces to work?

File: digits/scheduler.py [hunk 3]
```diff
@@ -110,37 +116,49 @@ def load_past_jobs(self):
...
         failed = 0
         loaded_jobs = []
-        for dir_name in sorted(os.listdir(config_value('jobs_dir'))):
-            if os.path.isdir(os.path.join(config_value('jobs_dir'), dir_name)):
-                exists = False
-
+        for workspace in sorted(os.listdir(config_value('jobs_dir'))):
+            if not self.workspace_jobs.has_key(workspace):
+                self.workspace_jobs[workspace] = []
+
```
Comment (lukeyeager) [inline digits/scheduler.py]: There's some extra whitespace here and elsewhere.
workspace-whitespace

**Comment by ahmedosman** [comment]
Re- Old jobs being parsed as work spaces, We assumed that no jobs existed before starting using work space. I'll update the scheduler to parse the jobs folder first and move the jobs in the default work space.

File: digits/dataset/images/classification/views.py [hunk 4]
```diff
@@ -314,18 +318,19 @@ def image_classification_dataset_create():
...
-            return flask.redirect(flask.url_for('datasets_show', job_id=job.id()))
+            return flask.redirect(flask.url_for('datasets_show', job_id=job.id())+'?workspace='+workspace)
 
     except:
         if job:
             scheduler.delete_job(job)
         raise
 
-def show(job):
+def show(job, *args):
```
Comment (jmancewicz) [inline digits/dataset/images/classification/views.py]: Maybe call the second parameter `workspace` rather than `args`, because that's the only way it's called.

File: digits/model/images/classification/views.py [hunk 5]
```diff
@@ -215,18 +220,18 @@ def image_classification_model_create():
...
-            return flask.redirect(flask.url_for('models_show', job_id=job.id()))
+            return flask.redirect(flask.url_for('models_show', job_id=job.id())+'?workspace='+workspace)
 
     except:
         if job:
             scheduler.delete_job(job)
         raise
 
-def show(job):
+def show(job, *args):
```
Comment (jmancewicz) [inline digits/model/images/classification/views.py]: Here too.

File: digits/views.py [hunk 6]
```diff
@@ -90,11 +99,16 @@ def home():
                 completed_models    = completed_models,
                 total_gpu_count     = len(scheduler.resources['gpus']),
                 remaining_gpu_count = sum(r.remaining() for r in scheduler.resources['gpus']),
+                workspace = workspace,
                 )
 
-def get_job_list(cls, running):
+def get_job_list(cls, running, *args):
```
Comment (jmancewicz) [inline digits/views.py]: And here.

File: digits/job.py [hunk 7]
```diff
@@ -23,14 +23,14 @@ class Job(StatusCls):
     SAVE_FILE = 'status.pickle'
 
     @classmethod
-    def load(cls, job_id):
+    def load(cls, path_job):
```
Comment (jmancewicz) [inline digits/job.py]: Just being picky, but maybe call this `job_path`.

File: digits/job.py [hunk 8]
```diff
@@ -43,22 +43,23 @@ def load(cls, job_id):
...
         # create a unique ID
+        super(Job, self).__init__()
         self._id = '%s-%s' % (time.strftime('%Y%m%d-%H%M%S'), os.urandom(2).encode('hex'))
-        self._dir = os.path.join(config_value('jobs_dir'), self._id)
+        self._dir = os.path.join(config_value('jobs_dir'), workspace, self._id)
         self._name = name
         self.pickver_job = PICKLE_VERSION
         self.tasks = []
         self.exception = None
-        self._notes = None
```
Comment (jmancewicz) [inline digits/job.py]: Did you mean to remove `self._notes`?

File: digits/model/tasks/caffe_train.py [hunk 9]
```diff
@@ -1184,6 +1184,66 @@ def get_layer_visualizations(self, net, layers='all'):
...
+                                        {
+                                            'name': str(top),
+                                            'type': 'Activation',
+                                            'mean': mean,
+                                            'stddev': std,
+                                            'histogram': hist,
+                                            'image_html': utils.image.embed_image_html(vis),
+                                        }
+                                    )
+                                added_activations.append(top)
```
Comment (jmancewicz) [inline digits/model/tasks/caffe_train.py]: This bit seems unrelated to the workspace concept.  Should this be a separate PR?

File: digits/templates/workspaces.html [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+						<div class="col-md-4">
+							<button id="submitbtn" name="submitbtn" class="btn btn-primary">Submit</button>
+						</div>
+					</div>
+				</form>
+			</div>
+		</div>
+
+	</div>	
+</body>
```
Comment (jmancewicz) [inline digits/templates/workspaces.html]: Can you change this to 4 space tabs?

File: digits/workspaces.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
+from urlparse import urlparse, parse_qs
+
+def get_workspace(url):
+	try:
+		workspace_name = parse_qs(urlparse(str(url)).query, keep_blank_values=True)["workspace"][0].encode('utf-8')
+		if workspace_name == '':
+			workspace_name = "Default"
+	except:
+		workspace_name = "Default"
+	return workspace_name
```
Comment (jmancewicz) [inline digits/workspaces.py]: Can you add the last newline?

**Comment by lukeyeager** [comment]
On second thought, before you go to all the work of moving job folders around, why do we need the workspaces to have their own sub-folders? Why can't they still all live in the same parent directory? That would simplify things.

**Comment by jmancewicz** [comment]
Then would all previous jobs without a workspace specified go into a global space?

**Comment by jmancewicz** [comment]
I had actually been planning on just adding a workspace item to the job. Other than that this seems very much like what I had intended to do.  For clarity, it make more sense to me for the argument order to be `workspace, job_id` rather than `job_id, workspace` because the workspace is like a namespace, and those read from left to right.

File: digits/model/tasks/caffe_train.py (hunk 9 above)
Comment (ahmedosman) [inline digits/model/tasks/caffe_train.py]: Yes, will remove it, that belongs to feature extraction code snippet, this will be part of a separate pull request .

File: digits/job.py (hunk 7 above)
Comment (ahmedosman) [inline digits/job.py]: :) Ok.

**Comment by ahmedosman** [comment]
@lukeyeager  We placed workspaces above the jobs in the hierarchy for the use case where we'd like to recall all the jobs that belongs to a specific workspace (for example when training a model from inside a workspace and we want to call only the data jobs relevant to that workspace), it's much easier to search only inside the $workspace_name folder than a folder contains all the jobs from different workspaces. @jmancewicz  by default the scheduler will copy all the jobs to the Default workspace (global workspace).

**Comment by ahmedosman** [comment]
@lukeyeager  Hi Luke I factored in the code review changes, can you look at/test commit 1f7e551, every time the server start, we scan folder inside $job directory, if a folder contains a "status.pickle" file then it is considered as a job and move to $job/Default which is the global workspace.

**Comment by jmancewicz** [comment]
Looks good so far.  I'll test it in the morning.

File: digits/model/tasks/caffe_train.py [hunk 12]
```diff
@@ -1184,8 +1184,8 @@ def get_layer_visualizations(self, net, layers='all'):
                                         }
                                     )
                             added_activations.append(top)
-            else:
-                raise NotImplementedError
+        else:
+            raise NotImplementedError
```
Comment (lukeyeager) [inline digits/model/tasks/caffe_train.py]: Why the whitespace change here?

File: digits/scheduler.py [hunk 13]
```diff
@@ -110,43 +119,62 @@ def load_past_jobs(self):
...
+                                    print '\t%s' % e
+                                else:
+                                    print 'Caught %s while loading job "%s"' % (type(e).__name__, dir_name)
 
         # add DatasetJobs
         for job in loaded_jobs:
             if isinstance(job, DatasetJob):
                 self.jobs.append(job)
-
+
```
Comment (lukeyeager) [inline digits/scheduler.py]: Extraneous whitespace.

**Comment by ahmedosman** [comment]
Hi @lukeyeager , apparently last commit broke a test (cc @DESHRAJ  ) which was basically removing spaces. Any idea why?

File: digits/views.py [hunk 14]
```diff
@@ -19,6 +19,11 @@
 from digits.utils import errors
 from digits.utils.routing import request_wants_json
 from digits.log import logger
+from flask import Flask, request, render_template
+from workspaces import get_workspace
+import shutil
+import os
+from boto.ec2.autoscale import scheduled
```
Comment (lukeyeager) [inline digits/views.py]: What is this for? If you're going to add a new library, you have to add it to the list of requirements. But it doesn't look like you're actually using this import ...

**Comment by lukeyeager** [comment]
Looking at the Travis logs, you've got issues with trying to import from `boto` and with your Job constructor.
https://travis-ci.org/NVIDIA/DIGITS/builds/94274106

Sorry for the slow response on this PR in general - we're busy finishing up the v3.0 release. Once that is finished, we should be able to give this PR the attention it deserves and get it merged.

**Comment by lukeyeager** [comment]
Hey @ahmedosman, please take a look at what I'm proposing for password-less "authentication" in #463. How would you feel about pivoting this PR to filter by username rather than workspace? I'm open to pushback if you feel like workspaces are a better choice.

**Comment by ahmedosman** [comment]
Hey @lukeyeager  so the proposed passwordlesss authentication seems to be very similar to workspaces with users instead of workspaces. Jobs that are created within a workspace are only visible within the workspace they were created in, so they can't be cancelled/deleted from outside a workspace. 

We added user authentication for digits, however the user authentication code was written in django and lives in the cloudcv code base . A user based system is certainly a good feature, however I feel making it passwordless makes it very similar to workspaces and defeat the purpose of having a user based system. If you are going to adopt a user based system in DIGITS, I recommend considering a password protected user system with a feature to link social accounts (such as google drive, dropbox, amazon S3) to a users account so that users can import dataset from the cloud (you will need to have a password protected account on digits in this case).

Sorry for the lag in getting back to the thread, @DESHRAJ and myself are in an exams period and very occupied by them. We have another PR to make for feature exportation (.npy and .mat files) and filter visualization that's blocked by this. Could you let us know what's blocking this PR ?

File: digits/workspaces.py [hunk 15]
```diff
@@ -0,0 +1,11 @@
+from urlparse import urlparse, parse_qs
+
+def get_workspace(url):
+	try:
+		workspace_name = parse_qs(urlparse(str(url)).query, keep_blank_values=True)["workspace"][0].encode('utf-8')
```
Comment (lukeyeager) [inline digits/workspaces.py]: 1. Rather than parsing the URL, let's use `flask.request.args`: https://github.com/NVIDIA/DIGITS/blob/f9ad5ced48/digits/utils/routing.py#L40-L41
2. Let's move this into `utils.routing` rather than creating a new file
3. Please use 4 spaces for indentation instead of tabs (here and elsewhere): https://www.python.org/dev/peps/pep-0008/#tabs-or-spaces

**Comment by lukeyeager** [comment]
> the proposed passwordlesss authentication seems to be very similar to workspaces with users instead of workspaces

Yes, that's true. I'll bring it up with the team and see which direction we want to go in.

> We added user authentication for digits, however the user authentication code was written in django and lives in the cloudcv code base . A user based system is certainly a good feature, however I feel making it passwordless makes it very similar to workspaces and defeat the purpose of having a user based system.

How complicated did your setup get? Did you have permissions on a per-workspace basis where only certain users could make changes to a workspace? Could any user switch to any workspace and then create/delete/edit data? I'm just curious.

> Sorry for the lag in getting back to the thread, @DESHRAJ and myself are in an exams period and very occupied by them. We have another PR to make for feature exportation (.npy and .mat files) and filter visualization that's blocked by this. Could you let us know what's blocking this PR?

No problem about the slow response! I was slow in responding as well due to our code freeze for v3.0 and due to all the packaging work that needed to get done. Now that we are unblocked for v3.1 development, I would love to get you unblocked so that you can continue contributing!

That being said, there are several things that still need to be addressed before I could merge this:

1. See the `boto` comment above - it's causing the TravisCI builds to fail, and requires me to either delete the line or install an unnecessary package on my machine in order to get your code to run.
2. As mentioned by Joe, we use 4 spaces for indentation instead of tabs.
3. This PR is not merge-able at the moment, probably because I merged #465. Sorry about that, I didn't realize that it would conflict with your PR!

File: digits/templates/models/images/classification/new.html [hunk 16]
```diff
@@ -74,7 +74,7 @@
          python_layer_file_state();
 
 $("#dataset").change(function() {
-        $.ajax("/datasets/images/classification/summary?job_id=" + $(this).val(),
+        $.ajax("/datasets/images/classification/summary?job_id=" + $(this).val()?workspace={{ workspace }},
```
Comment (lukeyeager) [inline digits/templates/models/images/classification/new.html]: Javascript error here: `Uncaught SyntaxError: Unexpected token ,`

File: digits/model/views.py [hunk 17]
```diff
@@ -102,9 +102,9 @@ def models_show(job_id):
         return flask.jsonify(job.json_dict(True))
     else:
         if isinstance(job, model_images.ImageClassificationModelJob):
-            return model_images.classification.views.show(job)
+            return model_images.classification.views.show(job, workspace)
         elif isinstance(job, model_images.GenericImageModelJob):
-            return model_images.generic.views.show(job)
+            return model_images.generic.views.show(job, workspace)
```
Comment (lukeyeager) [inline digits/model/views.py]: Error here: `NameError: global name 'workspace' is not defined`

File: digits/job.py [hunk 18]
```diff
@@ -43,22 +43,24 @@ def load(cls, job_id):
                     task.detect_snapshots()
             return job
 
-    def __init__(self, name):
+    def get_path(self):
+        return self._dir
+    
+    def __init__(self, name, workspace):
```
Comment (lukeyeager) [inline digits/job.py]: This is causing a bug when trying to create a new ImageClassificationModelJob:
```python
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/lib/python2.7/dist-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/lyeager/digits/digits/model/images/classification/views.py", line 100, in image_classification_model_create
    dataset_id  = datasetJob.id(),
  File "/home/lyeager/digits/digits/model/images/classification/job.py", line 18, in __init__
    super(ImageClassificationModelJob, self).__init__(**kwargs)
  File "/home/lyeager/digits/digits/model/images/job.py", line 16, in __init__
    super(ImageModelJob, self).__init__(**kwargs)
  File "/home/lyeager/digits/digits/model/job.py", line 21, in __init__
    super(ModelJob, self).__init__(**kwargs)
TypeError: __init__() takes exactly 3 arguments (2 given)
```

File: digits/templates/home.html [hunk 19]
```diff
@@ -30,6 +30,44 @@
...
+    return false;
+    }
+</script>
+<br>
+        <div class="row">
+        <div class="col-sm-8">
+
+            <form action="/workspace/edit/{{ workspace }}" method="post">
+            <div id="show-workspace-name" class="col-sm-12">
+                <h2 style="display:inline;">{{ workspace }}</h2>
```
Comment (lukeyeager) [inline digits/templates/home.html]: Can we find a way to improve the look of this? A huge "Default" without any context at the top of the page is a bit odd and confusing.

Maybe it would be clearer if (1) the current workspace, (2) the link to the list of workspaces and (3) the button to delete the workspace were all next to each other? Maybe we should move more of this up into the navbar? You can take a look at what I did for the login PR for inspiration if it's helpful - #463.

**Comment by DESHRAJ** [comment]
Hi @lukeyeager (cc @ahmedosman ), I have fixed the bugs suggested by you. Also, I will change the home template display for the workspaces in next two days. Rest seeems working fine as all the tests are passing. Let me know if any other changes are required. :)

**Comment by lukeyeager** [comment]
@DESHRAJ the pull request still cannot be merged. See this message at the bottom?

> This branch has conflicts that must be resolved

**Comment by lukeyeager** [comment]
Is this still a pull request branch? Is this your main development branch?

Integrating workspaces is more complicated now that we've merged https://github.com/NVIDIA/DIGITS/pull/463 - how would the two systems interact? Would you be able to filter by user and/or workspace?
